File: Day21.hs
import Control.Applicative (liftA2)
import Text.Read (readMaybe)

data Tree = Leaf String Int
          | Node Char Tree Tree

build :: String -> [String] -> Tree
build key lines = case readMaybe (drop 6 line) :: Maybe Int
    of (Just val) -> Leaf key val
       Nothing -> Node op (build lhs lines) (build rhs lines)
  where line = head $ filter ((==key) . take 4) lines -- O(n) :(
        [_, lhs, [op], rhs] = words line

fromChar :: Char -> Int -> Int -> Int
fromChar '+' = (+); fromChar '-' = (-)
fromChar '*' = (*); fromChar '/' = div

-- Part1: just evaluate the tree
eval :: Tree -> Int
eval (Leaf _ val) = val
eval (Node op lhs rhs) = (fromChar op) (eval lhs) (eval rhs)

-- Given an operation, one of the operands and
-- the goal, returns the value of the other operand
findL, findR :: Char -> Int -> Int -> Int
findL '+' = (-); findL '-' = (+)
findL '*' = div; findL '/' = (*)
findR '+' = flip (-); findR '-' = (-)
findR '*' = flip div; findR '/' = div

-- Part2: finds the value of "humn" leaf such that both subtrees eval to the same value
solve :: Tree -> Int
solve (Node _ lhs rhs) = case (eval2 lhs, eval2 rhs)
    of (Just lhsRes, Nothing) -> find lhsRes rhs
       (Nothing, Just rhsRes) -> find rhsRes lhs
  where -- Returns Nothing if "humn" is among the leaves, equivalent to (Just . eval) otherwise
        eval2 :: Tree -> Maybe Int
        eval2 (Leaf "humn" _) = Nothing
        eval2 (Leaf _ val) = Just val
        eval2 (Node op lhs rhs) = (liftA2 $ fromChar op) (eval2 lhs) (eval2 rhs)
        -- Given a goal & tree with "humn" as leaf, finds the value of this leaf
        find :: Int -> Tree -> Int
        find goal (Leaf "humn" _) = goal
        find goal (Node op lhs rhs) = case (eval2 lhs, eval2 rhs)
            of (Just lhsRes, Nothing) -> find (findR op lhsRes goal) rhs
               (Nothing, Just rhsRes) -> find (findL op goal rhsRes) lhs

main :: IO ()
main = do
    t1 <- build "root" . lines <$> readFile "input/21test.txt"
    t2 <- build "root" . lines <$> readFile "input/21full.txt"
    print $ (eval t1, solve t1) -- (152,301)
    print $ (eval t2, solve t2) -- (93813115694560,3910938071092)

File: Day21.cpp
#include "Lines.h"
#include <fmt/core.h>
#include <ranges>
#include <variant>
#include <charconv>
#include <map>
#include <algorithm>

struct Node {
	std::string lhs, rhs; // Operand keys
	char op; // Operation
};
using Expr = std::map<std::string, std::variant<int64_t, Node>>;

struct EvalVisitor {
	const Expr& e;
	int64_t operator()(int64_t x) const { return x; }
	int64_t operator()(const Node& node) const {
		const int64_t lhs = std::visit(*this, e.at(node.lhs));
		const int64_t rhs = std::visit(*this, e.at(node.rhs));
		switch (node.op) {
		case '+': return lhs + rhs;
		case '-': return lhs - rhs;
		case '*': return lhs * rhs;
		case '/': return lhs / rhs;
		}
		assert(false); return 0;
	}
};

struct SolveVisitor {
	const Expr& e;
	int64_t operator()(int64_t x) const { return x; }
	int64_t operator()(const Node& node) const {
		if (node.lhs == "humn" || node.rhs == "humn") {
			return -1;
		} else {
			const int64_t lhs = std::visit(*this, e.at(node.lhs));
			const int64_t rhs = std::visit(*this, e.at(node.rhs));
			if (lhs == -1 || rhs == -1) { return -1; } // Propagate upwards
			switch (node.op) {
			case '+': return lhs + rhs;
			case '-': return lhs - rhs;
			case '*': return lhs * rhs;
			case '/': return lhs / rhs;
			}
			assert(false); return 0;
		}
	}
};
// Given an operation, its result and one of the operands, finds the other operand
int64_t find(char op, int64_t rhs, int64_t goal) { // Only the right operand is known
	switch (op) {
	case '+': return goal - rhs;
	case '-': return goal + rhs;
	case '*': return goal / rhs;
	case '/': return goal * rhs;
	}
	std::terminate();
}
int64_t find(int64_t lhs, char op, int64_t goal) { // Only the left operand is known
	switch (op) {
	case '+': return goal - lhs;
	case '-': return lhs - goal;
	case '*': return goal / lhs;
	case '/': return lhs / goal;
	}
	std::terminate();
}

int64_t solve(const Expr& e, const std::string& key, const int64_t goal) {
	if (key == "humn") {
		return goal;
	} else {
		assert(e.contains(key) && std::holds_alternative<Node>(e.at(key)));
		const auto& [lhs, rhs, op] = std::get<Node>(e.at(key));
		const int64_t lhsRes = (lhs == "humn" ? -1 : std::visit(SolveVisitor{ e }, e.at(lhs)));
		const int64_t rhsRes = (rhs == "humn" ? -1 : std::visit(SolveVisitor{ e }, e.at(rhs)));
		assert((lhsRes == -1) ^ (rhsRes == -1));
		return (lhsRes == -1
			? solve(e, lhs, find(op, rhsRes, goal))
			: solve(e, rhs, find(lhsRes, op, goal)));
	}
}

void day21(const char* filename) {
	Expr e;
	for (const std::string& line : lines(filename)) {
		const std::string key(line, 0, 4);
		int64_t val = 0;
		const auto [ptr, ec] = std::from_chars(line.data() + 6, line.data() + line.size(), val);
		if (ptr != line.data() + 6) {
			e[key] = val;
		} else {
			e[key] = Node{ std::string(line, 6, 4),std::string(line, 13, 4),line[11] };
		}
	}
	const int64_t part1 = std::visit(EvalVisitor{ e }, e.at("root"));
	// Ignore root operation, it's supposed to be = after all
	const auto& [lhs, rhs, _] = std::get<Node>(e.at("root"));
	const int64_t lhsRes = std::visit(SolveVisitor{ e }, e.at(lhs));
	const int64_t rhsRes = std::visit(SolveVisitor{ e }, e.at(rhs));
	assert((lhsRes == -1) ^ (rhsRes == -1));
	const int64_t part2 = (lhsRes == -1 ? solve(e, lhs, rhsRes) : solve(e, rhs, lhsRes));
	fmt::print("{} {}\n", part1, part2);
}

int main21() {
	day21("input/21test.txt"); // 152 301
	day21("input/21full.txt"); // 93813115694560 3910938071092
	return 0;
}
